### PR TITLE
Special character splatting

### PR DESCRIPTION
**Pull request by Ajay-MS**

### CODE REVIEW

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 1]
```diff
@@ -81,22 +75,73 @@ function GetScriptToRun
...
+            targetMethod=$targetMethod 
+            serverName=$serverName
+            databaseName=$databaseName
+            authscheme=$authscheme 
+            sqlServerCredentials="`$sqlServerCredentials"
+            connectionString=$connectionString 
+            publishProfile=$publishProfile 
+            additionalArguments=$additionalArguments
+        }
+        $jsonSqlSplatArguments = $sqlSplatArguments | ConvertTo-Json
```
Comment (chshrikh) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: will be good if we can place this in try catch and print the error in catch block

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 2]
```diff
@@ -81,22 +75,73 @@ function GetScriptToRun
...
+            taskType=$taskType
+            sqlFile=$sqlFile
+            inlineSql=$inlineSql
+            serverName=$serverName
+            databaseName=$databaseName
+            authscheme=$authscheme
+            sqlServerCredentials="`$sqlServerCredentials"
+            additionalArguments=$additionalArguments
+        }
+        $jsonSqlSplatArguments = $sqlSplatArguments | ConvertTo-Json
```
Comment (chshrikh) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: same as above use try catch

File: TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1 [hunk 3]
```diff
@@ -48,25 +48,49 @@ function Execute-SqlQueryDeployment
...
+        $spaltArguments = @{
+            ServerInstance=$serverName
+            Database=$databaseName
+            InputFile=$sqlFile
+        }
+
+        # Process Additional arguments 
+        if($additionalArguments) 
+        {
+            $additionArgsArray = $additionalArguments.Split()
```
Comment (chshrikh) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1]: you may need to trim each arg and value. Please reffer this  code which does the command line argument parsing same as you are doing here - https://github.com/Microsoft/vsts-agent/blob/f7e9a7ee42d0c30f0d31d449bb51d95bd6e9e0fb/src/Microsoft.VisualStudio.Services.Agent/CommandLineParser.cs

Comment (chshrikh) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1]: ex input -    --url  "http://tfs" --username "UserName" --password "rtrtr"

File: TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1 [hunk 4]
```diff
@@ -48,25 +48,49 @@ function Execute-SqlQueryDeployment
...
-        $commandToLog = $scriptArgument
+        $spaltArgumentsToLog = $spaltArguments
+        $spaltArgumentsToLogJson = $spaltArgumentsToLog | ConvertTo-Json
+        Write-Verbose "Arguments : $spaltArgumentsToLogJson"
 
         if($authscheme -eq "sqlServerAuthentication")
         {
             if($sqlServerCredentials)
             {
                 $sqlUsername = $sqlServerCredentials.GetNetworkCredential().username
```
Comment (chshrikh) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1]: GetNetworkCredential().username will remove the @ from user name

File: TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1 (hunk 3 above)
Comment (chshrikh) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1]: please try to have the unit test for parsing logic

File: TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1 [hunk 5]
```diff
@@ -48,25 +48,65 @@ function Execute-SqlQueryDeployment
...
+
+        # Process Additional arguments 
+        if($additionalArguments) 
+        {
+            $args = $additionalArguments.Split()
+            $argScope = $null
+            foreach($arg in $args) {
+                $arg = $arg.Trim()
+                if($arg)
+                {
```
Comment (chshrikh) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1]: better to have test coverage for this

**Review by chshrikh — APPROVED** [review]

File: TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1 [hunk 6]
```diff
@@ -422,7 +422,7 @@ function Get-SqlPackageCmdArgs
         {
             if($sqlServerCredentials)
             {
-                $sqlUsername = $sqlServerCredentials.GetNetworkCredential().username
+                $sqlUsername = $sqlServerCredentials.UserName
```
Comment (mvvsubbu) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1]: Curious if GetNetworkCredentials().username not working, how GetNetworkCredentials().password working? If it is working why to change?

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 7]
```diff
@@ -66,12 +66,6 @@ function GetScriptToRun
         [string]$additionalArguments
     )
 
-    $connectionString = EscapeSpecialChars -str $connectionString
-    $sqlPassword = EscapeSpecialChars -str $sqlPassword
-    $additionalArguments = EscapeSpecialChars -str $additionalArguments
-    $serverName = EscapeSpecialChars -str $serverName
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: Remove EscapeSpecialChars function as well

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 [hunk 8]
```diff
@@ -81,22 +75,88 @@ function GetScriptToRun
...
+                serverName=$serverName
+                databaseName=$databaseName
+                authscheme=$authscheme 
+                sqlServerCredentials="`$sqlServerCredentials"
+                connectionString=$connectionString 
+                publishProfile=$publishProfile 
+                additionalArguments=$additionalArguments
+            }
+            $jsonSqlSplatArguments = $sqlSplatArguments | ConvertTo-Json
+            $remoteArgsInit = "`$remoteJsonSqlSplattedArgs = '$jsonSqlSplatArguments'
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: Can you explain why is this required? Did you test the task with additional arguments and for all the different input scenarios?

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/Main.ps1 [hunk 9]
```diff
@@ -33,4 +33,26 @@ if ($taskType -ne "dacpac")
...
+    adminUserName=$adminUserName 
+    adminPassword=$adminPassword 
+    winrmProtocol=$winrmProtocol 
+    testCertificate=$testCertificate 
+    dacpacFile=$dacpacFile 
+    targetMethod=$targetMethod 
+    serverName=$serverName 
+    databaseName=$databaseName 
+    authscheme=$authscheme 
+    sqlUsername=$sqlUsername
```
Comment (mvvsubbu) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/Main.ps1]: I see there is no change in Main function then why is this change?

File: TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1 [hunk 10]
```diff
@@ -477,8 +467,34 @@ function Execute-DacpacDeployment
     $sqlPackageArguments = Get-SqlPackageCmdArgs -dacpacFile $dacpacFile -targetMethod $targetMethod -serverName $serverName -databaseName $databaseName -authscheme $authscheme -sqlServerCredentials $sqlServerCredentials -connectionString $connectionString -publishProfile $publishProfile -additionalArguments $additionalArguments
     Write-Verbose -Verbose $sqlPackageArguments
     
-    $command = "`"$sqlPackage`" $sqlPackageArguments"
-    Write-Verbose "Executing command: $command"
-    RunCommand -command $command -failOnErr $true
+    Write-Verbose "Executing command: $sqlPackage $sqlPackageArguments"
+    Invoke-Command -FileName "$sqlPackage"  -Arguments $sqlPackageArguments
```
Comment (mvvsubbu) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1]: Let's not override build in function invoke-command instead name it as execute-sqlScript

File: TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1 (hunk 5 above)
Comment (mvvsubbu) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1]: I believe there is some regular expression in dtl sdk layer which is used to split the string command to dictionary of key value pairs. You can refer DTL SDK code for this and use the same logic which is already tested. 

You can refer code path of ScriptContent with arguments, you can take Chaitanya help to figure out where to start.

File: TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1 [hunk 11]
```diff
@@ -48,25 +48,65 @@ function Execute-SqlQueryDeployment
...
+        $spaltArgumentsToLog = $spaltArguments
+        $spaltArgumentsToLogJson = $spaltArgumentsToLog | ConvertTo-Json
+        Write-Verbose "Arguments : $spaltArgumentsToLogJson"
 
         if($authscheme -eq "sqlServerAuthentication")
         {
             if($sqlServerCredentials)
             {
-                $sqlUsername = $sqlServerCredentials.GetNetworkCredential().username
+                $sqlUsername = $sqlServerCredentials.Username
```
Comment (mvvsubbu) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1]: same as above?

File: TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1 [hunk 12]
```diff
@@ -448,16 +448,6 @@ function Get-SqlPackageCmdArgs
     return $sqlPkgCmdArgs
 }
 
-function Escape-SpecialChars
-{
-    param(
-        [string]$str
```
Comment (mvvsubbu) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1]: You are removing this function from utility module what is the impact on the consuming module?

**Review by mvvsubbu — CHANGES_REQUESTED** [review]
Fix comments

File: TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1 (hunk 6 above)
Comment (Ajay-MS) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1]: GetNetworkCredentials().username is removing @'s if more than one @ are there

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1 (hunk 8 above)
Comment (Ajay-MS) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/DeployToSqlServer.ps1]: We are using splatting to prevent expansion of special character, there is not way directly to pass objects to remote machine for execution, So we need to serialize it to json and then on remote machine we are expanding it.

File: Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/Main.ps1 (hunk 9 above)
Comment (Ajay-MS) [inline Extensions/IISWebAppDeploy/Src/Tasks/SqlDacpacDeploy/Main.ps1]: Arguments are being passed as splat arguments now.

File: TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1 (hunk 12 above)
Comment (Ajay-MS) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlPackageOnTargetMachines.ps1]: It will continue to consume module being mentioned there which was being uploaded there. When we change version there at that time we need to remove this method there as well. Should I create PR for that repo also?

File: TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1 (hunk 5 above)
Comment (chshrikh) [inline TaskModules/powershell/TaskModuleSqlUtility/SqlQueryOnTargetMachines.ps1]: it is done with "Get-ParsedSessionVariables" which is commandlet not part of dtl-sdk

**Review by mvvsubbu — APPROVED** [review]
As discussed offline go with this logic and fix renaming of the function and commit

**Review by mvvsubbu — APPROVED** [review]
